version: "3.9"

services:
  postgres:
    image: postgres:17.5
    container_name: fastapi_postgres_dev
    ports:
      - "5433:5432"
    env_file:
      - .env.development
    environment:
      - POSTGRES_PASSWORD=${DATABASE_PASSWORD}
      - POSTGRES_DB=${DATABASE_NAME}
      - POSTGRES_USER=${DATABASE_USERNAME}
    volumes:
      - postgres_dev_data:/var/lib/postgresql/data
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 5s
      timeout: 5s
      retries: 5
    networks:
      - fastapi_network

  api:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: fastapi_api_dev
    depends_on:
      postgres:
        condition: service_healthy
    ports:
      - "8000:8000"
    volumes:
      - ./:/usr/src/app
      - /usr/src/app/__pycache__ # Exclude __pycache__
    command: bash -c "alembic upgrade head && uvicorn app.main:app --host 0.0.0.0 --port 8000 --reload"
    env_file:
      - .env.development
    environment:
      - DATABASE_HOSTNAME=postgres
      - DATABASE_PORT=5432
    networks:
      - fastapi_network

volumes:
  postgres_dev_data:

networks:
  fastapi_network:
    driver: bridge
